sycl/include/CL/sycl/ONEAPI/intel_matrix/matrix-amx.hpp

```diff
@@ -0,0 +1,439 @@
+  // tcols: Num of tiles in column.
+  static constexpr size_t tcols =
+      (NUM_COLS * sizeof(T) / 4 + TILE_SZ - 1) / TILE_SZ;
+  static constexpr size_t size = trows * tcols * TILE_SZ * TILE_SZ * 4;
+  // stride is aligned to T instead of int8
+  static constexpr size_t stride = tcols * TILE_SZ * 4 / sizeof(T);
+  _tile1024i tile;
+  // FIXME: isSmall is useful?
+  static constexpr bool isSmall = true;
+  matrix_layout layout;
+  // We do zero-padding for matrix whose size is not fitted into tiles in ctor.
+  joint_matrix(Group sg) {}
+};
+
+template <typename T> class submatrix {
+public:
+  _tile1024i tile;
+  short rows, cols;
+};
+template <typename Group, typename T, size_t JM_NUM_ROWS, size_t JM_NUM_COLS,
+          matrix_layout JM_LAYOUT>
+__attribute__((always_inline)) static typename std::enable_if<
+    (JM_NUM_ROWS > TILE_SZ) || (JM_NUM_COLS * sizeof(T) / 4 > TILE_SZ),
+    void>::type
+submatrix_load(detail::submatrix<T> &sub_m,
+               joint_matrix<Group, T, JM_NUM_ROWS, JM_NUM_COLS, JM_LAYOUT> jm,
+               uint32_t row, uint32_t col, size_t stride, matrix_layout layout,
+               bool shouldreload) {
+  uint32_t offset = (row * stride + col);
+  T *ptr = (T *)jm.raw_storage;
+  ptr += offset;
+  stride *= sizeof(T);
+  sub_m.rows = TILE_SZ;
+  sub_m.cols = TILE_SZ * 4;
+  sub_m.tile =
+      tileloadd64_internal(sub_m.rows, sub_m.cols, (char *)ptr, stride);
+}
+
+template <typename Group, typename T, size_t JM_NUM_ROWS, size_t JM_NUM_COLS,
+          matrix_layout JM_LAYOUT>
+__attribute__((always_inline)) static typename std::enable_if<
+    (JM_NUM_ROWS <= TILE_SZ) && (JM_NUM_COLS * sizeof(T) / 4 <= TILE_SZ),
+    void>::type
+submatrix_load(detail::submatrix<T> &sub_m,
+               joint_matrix<Group, T, JM_NUM_ROWS, JM_NUM_COLS, JM_LAYOUT> &jm,
+               uint32_t row, uint32_t col, size_t stride, matrix_layout layout,
+               bool shouldreload) {
+  if (shouldreload) {
+    // Force sub_m.tile's shape to be TILE_SZ * TILE_SZ * 4
+    int8_t NewjmC[TILE_SZ * TILE_SZ * 4];
+    tilestored64_internal(JM_NUM_ROWS, JM_NUM_COLS * sizeof(T), (char *)NewjmC,
+                          TILE_SZ * 4, jm.tile);
+    sub_m.rows = TILE_SZ;
+    sub_m.cols = TILE_SZ * 4;
+    sub_m.tile = tileloadd64_internal(sub_m.rows, sub_m.cols, (char *)NewjmC,
+                                      TILE_SZ * 4);
+    return;
+  }
+  sub_m.rows = JM_NUM_ROWS;
+  sub_m.cols = JM_NUM_COLS * sizeof(T);
+  sub_m.tile = jm.tile;
+}
+
+// This handles cases where T1 is int8, T2 is int32.
+template <typename T1, typename T2>
+__attribute__((always_inline)) static
+    typename std::enable_if<std::is_same<T1, int8_t>::value &&
+                                std::is_same<T2, int32_t>::value,
+                            void>::type
+    submatrix_mad(detail::submatrix<T1> &sub_ma, detail::submatrix<T1> &sub_mb,
+                  detail::submatrix<T2> &sub_mc) {
```

Comment:
In this case, it's better to use a function overload instead of template instantiation.
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#t144-dont-specialize-function-templates
```suggestion
// This handles cases where T1 is int8, T2 is int32.
__attribute__((always_inline)) static void 
submatrix_mad(detail::submatrix<int8_t> &sub_ma, detail::submatrix<int8_t> &sub_mb,
                  detail::submatrix<int32_t> &sub_mc) {
```
Please, apply to other function template specializations.